service/healthz/service.go

```diff
@@ -0,0 +1,97 @@
+package healthz
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	microerror "github.com/giantswarm/microkit/error"
+	micrologger "github.com/giantswarm/microkit/logger"
+	vaultapi "github.com/hashicorp/vault/api"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	healthCheckRequests = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "health_check_request_total",
+			Help: "Number of health check requests",
+		},
+		[]string{"success"},
+	)
+	healthCheckRequestTime = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "health_check_request_milliseconds",
+		Help: "Time taken to respond to health check, in milliseconds",
+	})
+)
+
+// Config represents the configuration used to create a healthz service.
+type Config struct {
+	// Dependencies.
+	Logger      micrologger.Logger
+	VaultClient *vaultapi.Client
+}
+
+// DefaultConfig provides a default configuration to create a new healthz
+// service by best effort.
+func DefaultConfig() Config {
+	return Config{
+		// Dependencies.
+		VaultClient: nil,
+		Logger:      nil,
+	}
+}
+
+// New creates a new configured healthz service.
+func New(config Config) (*Service, error) {
+	// Dependencies.
+	if config.Logger == nil {
+		return nil, microerror.MaskAnyf(invalidConfigError, "logger must not be empty")
+	}
+	if config.VaultClient == nil {
+		return nil, microerror.MaskAnyf(invalidConfigError, "vault client must not be empty")
+	}
+
+	newService := &Service{
+		// Dependencies.
+		vaultClient: config.VaultClient,
+		logger:      config.Logger,
+
+		// Internals.
+		bootOnce: sync.Once{},
+	}
+
+	return newService, nil
+}
+
+// Service implements the healthz service interface.
+type Service struct {
+	// Dependencies.
+	vaultClient *vaultapi.Client
+	logger      micrologger.Logger
+
+	// Internals.
+	bootOnce sync.Once
+}
+
+// Check implements the health check which lists the mounts for the sys backend.
+// This checks that the operator can connect to the Vault API and the token is
+// valid.
+func (s *Service) Check(ctx context.Context, request Request) (*Response, error) {
```

Comment:
Check method is where the custom logic is. I chose to list the sys mounts as the API call as it shouldn't have any negative effects on the Vault server while still testing the token.

As mentioned in the PR desc I'd like to switch the new Vault health point once its released and we are up to date.

https://github.com/hashicorp/vault/blob/master/api/sys_health.go